R/choices_labeled.R

```diff
@@ -171,35 +175,48 @@ choices_labeled <- function(choices, labels, subset = NULL, types = NULL) {
 #' })
 #' @export
 #'
-variable_choices <- function(data, subset = NULL, fill = FALSE, key = NULL) {
+variable_choices <- function(data = "all", subset = function(data) names(data), fill = FALSE, key = NULL) {
   checkmate::assert(
     checkmate::check_character(subset, null.ok = TRUE, any.missing = FALSE),
-    checkmate::check_function(subset)
+    checkmate::check_function(subset, args = "data")
   )
   checkmate::assert_flag(fill)
   checkmate::assert_character(key, null.ok = TRUE, any.missing = FALSE)
-
-  UseMethod("variable_choices")
+  if (is.character(data)) {
+    variable_choices_delayed(data = data, subset = subset, fill = fill, key = key)
+  } else {
+    variable_choices_data_frame(data = data, subset = subset, fill = fill, key = key)
+  }
 }
 
-#' @rdname variable_choices
-#' @export
-variable_choices.character <- function(data, subset = NULL, fill = FALSE, key = NULL) {
+#' @keywords internal
+variable_choices_delayed <- function(data, subset = function(data) names(data), fill = FALSE, key = NULL) {
+  checkmate::assert_string(data)
   structure(list(data = data, subset = subset, key = key),
     class = c("delayed_variable_choices", "delayed_data", "choices_labeled")
   )
 }
 
-#' @rdname variable_choices
-#' @export
-variable_choices.data.frame <- function(data, subset = NULL, fill = TRUE, key = NULL) {
+#' @keywords internal
+variable_choices_data_frame <- function(data, subset = function(data) names(data), fill = TRUE, key = NULL) {
+  checkmate::assert_data_frame(data, min.cols = 1)
   checkmate::assert(
     checkmate::check_character(subset, null.ok = TRUE),
     checkmate::check_function(subset, null.ok = TRUE)
   )
 
   if (is.function(subset)) {
-    subset <- resolve_delayed_expr(subset, ds = data, is_value_choices = FALSE)
+    subset <- subset(data)
+    if (
+      !checkmate::test_character(subset, any.missing = FALSE) ||
```

Comment:
Missing values are rejected by the caller.
```suggestion
      !checkmate::test_character(subset) ||
```